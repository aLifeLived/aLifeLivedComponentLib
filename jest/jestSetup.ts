/* eslint-disable no-undef */
jest.useFakeTimers({ legacyFakeTimers: true }); // Stops reference error from appearing
