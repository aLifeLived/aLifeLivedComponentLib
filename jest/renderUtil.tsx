import React from 'react';
import { ThemeProvider } from '../src/themes/ThemeProvider';
import { render as rtlRender } from '@testing-library/react-native';
import { library } from '@fortawesome/fontawesome-svg-core';
import { iconList } from '../src/themes/systemIcon/IconList';

library.add(iconList);

function render(ui: any, { ...renderOptions } = {}) {
  function Wrapper({ children }: { children: any }) {
    return <ThemeProvider>{children}</ThemeProvider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything for convenience
export * from '@testing-library/react-native';

// override render method
export { render };
