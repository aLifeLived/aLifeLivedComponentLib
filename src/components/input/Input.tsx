import React, { useState } from 'react';
import { TextInputProps } from 'react-native';
import { Theme } from '../../themes/theme';

// Hooks
import { useTheme } from '../../themes/hooks/useTheme';

// UI
import { Box } from '../../themes/box/Box';
import { Text } from '../../themes/text/Text';
import { Icon } from '../icon/Icon';
import {
  inputVariants,
  StyledInputProps,
  SystemInput,
} from '../../themes/systemInput/SystemInput';

interface InputProps extends TextInputProps {
  variant: keyof typeof inputVariants;
  label?: string;
  labelVariant?: keyof Omit<Theme['textVariants'], 'defaults'>;
}

const handleVariantState = (variant: keyof typeof inputVariants) => {
  if (variant === 'error') {
    return <ErrorIconAdornment />;
  } else if (variant === 'success') {
    return <SuccessIconAdornment />;
  } else {
    return null;
  }
};

const handleBorderColor = (variant: keyof typeof inputVariants) => {
  if (variant === 'error') {
    return 'backgroundError';
  } else if (variant === 'success') {
    return 'primary';
  } else {
    return 'borderColor';
  }
};

export const Input: React.FC<InputProps & StyledInputProps> = ({
  variant,
  label,
  labelVariant = 'caption',
  ...rest
}) => {
  const { colors } = useTheme();
  const endAdornmentComponent = handleVariantState(variant);
  const [isFocused, setIsFocused] = useState(false);
  const handleOnFocus = () => setIsFocused(true);
  const handleOnBlur = () => setIsFocused(false);

  return (
    <Box boxShadow="md">
      {label && (
        <Text marginBottom="s" variant={labelVariant}>
          {label}
        </Text>
      )}

      <Box flexDirection="row">
        <SystemInput
          variant={variant}
          flex={1}
          onFocus={handleOnFocus}
          onBlur={handleOnBlur}
          placeholderTextColor={colors.textContrastLow}
          borderColor={isFocused ? handleBorderColor(variant) : 'transparent'}
          {...rest}
        />
        {endAdornmentComponent ? (
          <InputIconAdornment>{endAdornmentComponent}</InputIconAdornment>
        ) : null}
      </Box>
    </Box>
  );
};

const InputIconAdornment: React.FC = React.memo(({ children }) => {
  const { spacing } = useTheme();
  return (
    <Box
      position="absolute"
      right={spacing.m}
      top={0}
      height="100%"
      alignItems="center"
      justifyContent="center"
      alignContent="center"
    >
      {children}
    </Box>
  );
});

const ErrorIconAdornment: React.FC = () => {
  return (
    <Icon
      icon="exclamation-circle"
      color="backgroundError"
      size={20}
      testID="O4G7U-error-icon"
    />
  );
};

const SuccessIconAdornment: React.FC = () => {
  return (
    <Icon
      icon="check-circle"
      color="primary"
      size={20}
      testID="E7P0M-success-icon"
    />
  );
};
