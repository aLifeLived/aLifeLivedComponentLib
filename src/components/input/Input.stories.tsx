import React from 'react';
import { storiesOf } from '@storybook/react-native';

// Component
import { Input } from './Input';
import { Box } from '../../themes/box/Box';

storiesOf('Input', module)
  .addDecorator(Story => Story())
  .add('Default', () => (
    <Box margin="s">
      <Input
        label="Hey this is a long label"
        variant="default"
        returnKeyType="default"
        placeholder="Hey there"
      />
    </Box>
  ))
  .add('No Label', () => (
    <Box margin="s">
      <Input
        variant="default"
        returnKeyType="default"
        placeholder="Hey there"
      />
    </Box>
  ));
