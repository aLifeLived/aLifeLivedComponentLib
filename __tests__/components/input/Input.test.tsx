import React from 'react';
import { render } from './../../../jest/renderUtil';

// Component
import { Input } from '../../../src/components/input/Input';

const mockOnPress = jest.fn();

const setup = ({ label, variant }) => {
  const utils = render(
    <Input
      placeholder="Test placeholder"
      label={label}
      variant={variant}
      onPressIn={mockOnPress}
    />
  );

  return {
    ...utils,
  };
};

describe('component >> Input', () => {
  describe('Given input has variant of default', () => {
    describe('Given input has label', () => {
      test('Should render input with label correctly', () => {
        const { toJSON } = setup({
          label: 'This is a test label',
          variant: 'default',
        });
        expect(toJSON()).toMatchSnapshot();
      });

      test('Should find label correctly', () => {
        const { getByText } = setup({
          label: 'This is a test label',
          variant: 'default',
        });
        expect(getByText('This is a test label')).toBeDefined();
      });
    });

    describe('Given input has NO label', () => {
      test('Should render input with label correctly', () => {
        const { toJSON } = setup({
          label: undefined,
          variant: 'default',
        });
        expect(toJSON()).toMatchSnapshot();
      });

      test('Should NOT find label correctly', () => {
        const { queryByText } = setup({
          label: undefined,
          variant: 'default',
        });
        expect(queryByText('This is a test label')).toBeNull();
      });
    });

    test('Should NOT find icon', () => {
      const { queryByTestId } = setup({
        label: 'This is a test label',
        variant: 'default',
      });
      expect(queryByTestId('E7P0M-success-icon')).toBeNull();
      expect(queryByTestId('O4G7U-error-icon')).toBeNull();
    });
  });

  describe('Given input has variant SUCCESS', () => {
    test('Should match snapshot correctly', () => {
      const { toJSON } = setup({
        label: 'Success variant',
        variant: 'success',
      });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find icon', () => {
      const { getByTestId } = setup({
        label: 'Success variant',
        variant: 'success',
      });
      expect(getByTestId('E7P0M-success-icon')).toBeDefined();
    });

    test('Should find label', () => {
      const { getByText } = setup({
        label: 'Success variant',
        variant: 'success',
      });
      expect(getByText('Success variant')).toBeDefined();
    });
  });

  describe('Given input has variant ERROR', () => {
    test('Should match snapshot correctly', () => {
      const { toJSON } = setup({
        label: 'Error variant',
        variant: 'error',
      });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find icon', () => {
      const { getByTestId } = setup({
        label: 'Error variant',
        variant: 'error',
      });
      expect(getByTestId('O4G7U-error-icon')).toBeDefined();
    });

    test('Should find label', () => {
      const { getByText } = setup({
        label: 'Error variant',
        variant: 'error',
      });
      expect(getByText('Error variant')).toBeDefined();
    });
  });

  describe('Given input has variant underline', () => {
    describe('Given input has label', () => {
      test('Should render input with label correctly', () => {
        const { toJSON } = setup({
          label: 'This is a test label',
          variant: 'underline',
        });
        expect(toJSON()).toMatchSnapshot();
      });

      test('Should find label correctly', () => {
        const { getByText } = setup({
          label: 'This is a test label',
          variant: 'underline',
        });
        expect(getByText('This is a test label')).toBeDefined();
      });
    });

    describe('Given input has NO label', () => {
      test('Should render input with label correctly', () => {
        const { toJSON } = setup({
          label: undefined,
          variant: 'underline',
        });
        expect(toJSON()).toMatchSnapshot();
      });

      test('Should NOT find label correctly', () => {
        const { queryByText } = setup({
          label: undefined,
          variant: 'underline',
        });
        expect(queryByText('This is a test label')).toBeDefined();
      });
    });
  });
});
