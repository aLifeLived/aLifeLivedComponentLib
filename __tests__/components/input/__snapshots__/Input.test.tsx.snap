// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component >> Input Given input has variant ERROR Should match snapshot correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
      Array [
        Object {
          "elevation": 2,
          "shadowColor": "#282828",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 10,
        },
      ],
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#7F8183",
          "display": "flex",
          "fontSize": 14,
          "lineHeight": 18,
          "marginBottom": 8,
        },
      ]
    }
  >
    Error variant
  </Text>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onPressIn={[MockFunction]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#D94E6928",
            "borderColor": "#D94E69",
            "borderRadius": 8,
            "borderWidth": 1,
            "color": "#282828",
            "flex": 1,
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 8,
            "paddingTop": 16,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
    <View
      style={
        Array [
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "display": "flex",
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "right": 16,
            "top": 0,
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={20}
        bbWidth={20}
        fill="#D94E69"
        focusable={false}
        height={20}
        meetOrSlice={0}
        minX={0}
        minY={0}
        secondaryFill={null}
        secondaryOpacity={0.4}
        size={20}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {},
            Object {
              "flex": 0,
              "height": 20,
              "width": 20,
            },
          ]
        }
        testID="O4G7U-error-icon"
        vbHeight={512}
        vbWidth={512}
        width={20}
      >
        <RNSVGGroup
          fill={4292431465}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
            fill={4292431465}
            fillOpacity={1}
            propList={
              Array [
                "fill",
                "fillOpacity",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
</View>
`;

exports[`component >> Input Given input has variant SUCCESS Should match snapshot correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
      Array [
        Object {
          "elevation": 2,
          "shadowColor": "#282828",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 10,
        },
      ],
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#7F8183",
          "display": "flex",
          "fontSize": 14,
          "lineHeight": 18,
          "marginBottom": 8,
        },
      ]
    }
  >
    Success variant
  </Text>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onPressIn={[MockFunction]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#67B6A849",
            "borderColor": "#67B6A8",
            "borderRadius": 8,
            "borderWidth": 1,
            "color": "#282828",
            "flex": 1,
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 8,
            "paddingTop": 16,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
    <View
      style={
        Array [
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "display": "flex",
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "right": 16,
            "top": 0,
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={20}
        bbWidth={20}
        fill="#67B6A8"
        focusable={false}
        height={20}
        meetOrSlice={0}
        minX={0}
        minY={0}
        secondaryFill={null}
        secondaryOpacity={0.4}
        size={20}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {},
            Object {
              "flex": 0,
              "height": 20,
              "width": 20,
            },
          ]
        }
        testID="E7P0M-success-icon"
        vbHeight={512}
        vbWidth={512}
        width={20}
      >
        <RNSVGGroup
          fill={4284987048}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
            fill={4284987048}
            fillOpacity={1}
            propList={
              Array [
                "fill",
                "fillOpacity",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
</View>
`;

exports[`component >> Input Given input has variant of default Given input has NO label Should render input with label correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
      Array [
        Object {
          "elevation": 2,
          "shadowColor": "#282828",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 10,
        },
      ],
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onPressIn={[MockFunction]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#2828281e",
            "borderColor": "#2828281e",
            "borderRadius": 8,
            "borderWidth": 1,
            "color": "#282828",
            "flex": 1,
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 8,
            "paddingTop": 16,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`component >> Input Given input has variant of default Given input has label Should render input with label correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
      Array [
        Object {
          "elevation": 2,
          "shadowColor": "#282828",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 10,
        },
      ],
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#7F8183",
          "display": "flex",
          "fontSize": 14,
          "lineHeight": 18,
          "marginBottom": 8,
        },
      ]
    }
  >
    This is a test label
  </Text>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onPressIn={[MockFunction]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#2828281e",
            "borderColor": "#2828281e",
            "borderRadius": 8,
            "borderWidth": 1,
            "color": "#282828",
            "flex": 1,
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 8,
            "paddingTop": 16,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`component >> Input Given input has variant underline Given input has NO label Should render input with label correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
      Array [
        Object {
          "elevation": 2,
          "shadowColor": "#282828",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 10,
        },
      ],
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onPressIn={[MockFunction]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "flex": 1,
            "paddingBottom": 8,
            "paddingTop": 8,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`component >> Input Given input has variant underline Given input has label Should render input with label correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
      Array [
        Object {
          "elevation": 2,
          "shadowColor": "#282828",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 10,
        },
      ],
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#7F8183",
          "display": "flex",
          "fontSize": 14,
          "lineHeight": 18,
          "marginBottom": 8,
        },
      ]
    }
  >
    This is a test label
  </Text>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onPressIn={[MockFunction]}
      placeholder="Test placeholder"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "flex": 1,
            "paddingBottom": 8,
            "paddingTop": 8,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
